eng/common/scripts/Generate-DocIndex.ps1

```diff
@@ -0,0 +1,149 @@
+    }
+    return $bomAwareString
+}
+
+function Get-PkgMetadata([Object[]]$metadata, [String]$artifactName, [Hashtable]$filterOptions) {
+    $filteredPkg = $metadata | ? { $_.Package -eq $artifactName }
+    if (!$filteredPkg) {
+        return
+    }
+    if ($filteredPkg.length -eq 1) {
+        return $filteredPkg[0]
+    }
+    foreach ($filter in $filterOptions.getEnumerator()) {
+        $filterKey = $filter.Key.Trim()
+        $filterValue = $filter.Value.Trim()
+        $filteredPkg = $filteredPkg | ? { $_.$filterKey -eq $filterValue }
+    }
+    if ($filteredPkg.length -ne 1) {
+        LogWarning "We fetch out $($filteredPkg.length) artifact(s) for the same package name $artifactName. "
+        LogWarning "Please check csv of Azure/azure-sdk/_data/release/latest repo if this is intended. "
+        return 
+    }
+    return $filteredPkg[0]
+}
+
+function Generate-DocIndex { 
+    Param (
+        [Parameter(Mandatory = $true)]  [Object[]]  $metadata,
+        [Parameter(Mandatory = $true)]  [Hashtable] $filterOptions,
+        [Parameter(Mandatory = $true)]  [String]    $blobStorageUrl,
+        [Parameter(Mandatory = $true)]  [String]    $blobDirectoryRegex,
+        [Parameter(Mandatory = $true)]  [String]    $titleRegex,
+        [Parameter(Mandatory = $false)] [String]    $packageNameRegex = '',
+        [Parameter(Mandatory = $false)] [String]    $packageNameReplacement = ''
+    )
+    Write-Verbose "Name Reccuring paths with variable names"
+    $DocFxTool = "${RepoRoot}/docfx/docfx.exe"
+    $DocOutDir = "${RepoRoot}/docfx_project"
+
+    Write-Verbose "Initializing Default DocFx Site..."
+    & $($DocFx) init -q -o "${DocOutDir}"
+
+    Write-Verbose "Copying template and configuration..."
+    New-Item -Path "${DocOutDir}" -Name "templates" -ItemType "directory" -Force
+    Copy-Item "${DocGenDir}/templates/*" -Destination "${DocOutDir}/templates" -Force -Recurse
+    Copy-Item "${DocGenDir}/docfx.json" -Destination "${DocOutDir}/" -Force
+    $YmlPath = "${DocOutDir}/api"
+    New-Item -Path $YmlPath -Name "toc.yml" -Force
+
+    Write-Verbose "Reading artifact from storage blob ..."
+    $pageToken = ""
+    # Used for sorting the toc display order
+    $orderSet = @{}
+    # This is a pagnation call as storage only return 5000 results as maximum.
+    Do {
+        $resp = ""
+        if (!$pageToken) {
+            # First page call.
+            $resp = Invoke-RestMethod -Method Get -Uri $blobStorageUrl
+        }
+        else {
+            # Next page call
+            $blobStorageUrlPageToken = $blobStorageUrl + "&marker=$pageToken"
+            $resp = Invoke-RestMethod -Method Get -Uri $blobStorageUrlPageToken
+        }
+        # Storage returns some weired encoded string at the begining of response which needs to cutoff before use.
+
+        # Convert to xml documents.
+        $xmlDoc = [xml](removeBomFromString $resp)
+        foreach ($elem in $xmlDoc.EnumerationResults.Blobs.BlobPrefix) {
```

Comment:
I think it would be a good idea to split the blob traversal and the generation of the yml/md files. This allows some flexibility in how we produce this list of packages. Perhaps this Generate-DocIndex function can take in a list of packages with exactly the data that is needed (i.e. a list of objects that has a package name and service name).

Reply:
Split the scripts to two parts: First, iterate through the blob storage and storage artifact name and service name into map.
Second, generate yml/md files for build docfx site